--- a/modules/nixos/main.py
+++ b/modules/nixos/main.py
@@ -647,6 +647,19 @@ def run():
     libcalamares.utils.host_env_process_output(
         ["cp", "/dev/stdin", config], None, cfg)

+    # Copying user provided configuraitons
+    dynamic_config = "/tmp/nixos-offline/nixos/configuration.nix"
+    iso_config = "/iso/nix-cfg/nixos/configuration.nix"
+    config_dest = os.path.join(root_mount_point, "etc/nixos/configuration.nix")
+
+    try:
+        if os.path.exists (dynamic_config):
+            subprocess.run(["sudo", "cp", dynamic_config, config_dest], check=True)
+        elif os.path.exists(iso_config):
+            subprocess.run(["sudo", "cp", iso_config, config_dest], check=True)
+    except subprocess.CalledProcessError as e:
+        return ("Installation failed to copy configuration files", str(e))
+
     status = _("Installing NixOS")
     libcalamares.job.setprogress(0.3)

@@ -663,6 +676,7 @@ def run():
         exit = proc.wait()
         if exit != 0:
             return (_("nixos-install failed"), _(output))
+
     except:
         return (_("nixos-install failed"), _("Installation failed to complete"))

