--- a/config/modules/welcome.conf
+++ b/config/modules/welcome.conf
@@ -82,7 +82,6 @@
         - storage
         - ram
         - power
-        - internet
         - screen
     # List conditions that **must** be satisfied (from the list
     # of conditions, above) for installation to proceed.
@@ -91,7 +90,6 @@
     required:
         - storage
         - ram
-        - internet
 
 # GeoIP checking
 #
